File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/DynamicProgramming/188_BestTimeBuySellStockIV.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-08-23 14:30:18
 */

c.. Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        size_t days = prices.size();
        __(k__0 || days<2){
            r_ 0;
        }

        __(k>prices.size()/2){
            r_ solve_quickly(prices);
        }

        vector<vector<int>> dp(k+1, vector<int>(days, 0));
        ___(int i=1; i<= k; i++){
            int max_buy = -prices[0];
            ___(int j=1; j<days; j++){
                dp[i][j] = m..(dp[i][j-1], max_buy+prices[j]);
                max_buy = m..(max_buy, dp[i-1][j-1]-prices[j]);
            }
        }
        r_ dp[k][days-1];
    }

    int solve_quickly(vector<int> &prices){
        int max_profit = 0;
        ___(int i=1; i<prices.size(); i++){
            int diff = prices[i] - prices[i-1];
            max_profit += diff > 0 ? diff : 0;
        }
        r_ max_profit;
    }
};

/*
1
[8,5,3]
2
[1,8,3,2,7,12,7,15,16,17]
4
[1,8,3,2,7,12,7,15,16,17]
*/

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/DynamicProgramming/188_BestTimeBuySellStockIV.py
# @Author: xuezaigds@gmail.com
# @Last Modified time: 2016-08-23 14:17:12
# -*- coding: utf-8 -*-


c.. Solution o..
    ___ maxProfit  k, prices
        __ n.. prices or k __ 0:
            r_ 0

        # There is no limit on transaction's times in fact.
        days_num = l..(prices)
        __ k / 2 > days_num:
            r_ self.quick_solve(prices)

        # dp[i][j] is the max profit for up to i transactions by days j.
        # 1 <= i <= K, 0 <= j < das_num.
        dp = [[0 ___ day __ r..(days_num)] ___ trans __ r..(k + 1)]
        ___ i __ r..(1, k + 1
            # tmpMax means the maximum profit if we've just buy i-th stock so far.
            tmp_max = -prices[0]
            ___ j __ r..(1, days_num
                dp[i][j] = m..(dp[i][j - 1], prices[j] + tmp_max)
                tmp_max = m..(tmp_max, dp[i - 1][j - 1] - prices[j])
        r_ dp[k][days_num - 1]

    # Get the max profits without the limit of transactions actually.
    ___ quick_solve  prices
        profits = 0
        ___ i __ r..(1, l..(prices)):
            price_diff = prices[i] - prices[i - 1]
            profits += price_diff __ price_diff > 0 else 0
        r_ profits

"""
1
[8,5,3]
2
[1,8,3,2,7,12,7,15,16,17]
4
[1,8,3,2,7,12,7,15,16,17]
"""
